Common/Source/Dialogs/dlgIGCProgress.cpp

```diff
@@ -0,0 +1,188 @@
+  hP.Release();
+
+  hP.Create(PEN_SOLID,NIBLSCALE(1),RGB_BLACK);
+  ohP = Surface.SelectObject(hP);
+  Surface.SelectObject(LK_HOLLOW_BRUSH);
+  InflateRect(&PrintAreaR, -NIBLSCALE(2), -NIBLSCALE(2));
+  Surface.Rectangle(PrintAreaR.left,PrintAreaR.top,PrintAreaR.right,PrintAreaR.bottom);
+
+  Surface.SetTextColor(RGB_WHITE);
+  Surface.SetBackgroundTransparent();
+
+  InflateRect(&PrintAreaR, -NIBLSCALE(2), -NIBLSCALE(2));
+  
+  const TCHAR* text = Sender->GetCaption();
+  Surface.DrawText(text, &PrintAreaR, DT_VCENTER|DT_SINGLELINE);
+
+  Surface.SelectObject(ohB);
+  Surface.SelectObject(ohP);
+  Surface.SelectObject(oldFont);
+
+}
+
+CallBackTableEntry_t IGCProgressCallBackTable[] = {
+    OnPaintCallbackEntry(OnIGCSplashPaint),
+    OnPaintCallbackEntry(OnIGCProgressPaint),
+    ClickNotifyCallbackEntry(OnIGCAbortClicked),
+    EndCallBackEntry()
+};
+
+class dlgIGCProgress final  {
+public:
+    dlgIGCProgress();
+    ~dlgIGCProgress();
+    
+    dlgIGCProgress( const dlgIGCProgress& ) = delete;
+    dlgIGCProgress& operator=( const dlgIGCProgress& ) = delete;
+
+};
+
+
+
+
+
+dlgIGCProgress::dlgIGCProgress() {};
+void dlgIGCProgressShowModal(void){
+    
+	bClose = false;
+
+    _WndForm = dlgLoadFromXML(IGCProgressCallBackTable, ScreenLandscape ? IDR_XML_IGC_PROGRESS_P : IDR_XML_IGC_PROGRESS_L);
+
+    LKASSERT(_WndForm);
+    if(_WndForm) {
+        WindowControl* wSplash = _WndForm->FindByName(TEXT("frmIGCSplash")); 
+        if(wSplash) {
+            wSplash->SetWidth(_WndForm->GetWidth());
+            wSplash->SetHeight(_WndForm->GetHeight());
+        }
+        WindowControl* wText = _WndForm->FindByName(TEXT("frmIGCText")); 
+        if(wText) {
+            wText->SetWidth(_WndForm->GetWidth());
+            wText->SetTop(_WndForm->GetHeight() - wText->GetHeight());
+        }
+        _WndForm->Show();
+        _WndForm->Redraw();
+        _WndForm->SetTimerNotify(200, OnIGCProgressTimer); // check for end of download every 200ms
+    }
+    _WndForm->SetTimerNotify(200, OnIGCProgressTimer); // check for end of download every 200ms
+    _WndForm->ShowModal();
+
+    delete _WndForm;
+    _WndForm = NULL;
+
+}
+
+dlgIGCProgress::~dlgIGCProgress() {
+    _WndForm->SetTimerNotify(0,NULL); // delete Timer
+    delete _WndForm; 
+}
+
+
+
+
+void CloseIGCProgressDialog() {
+    bClose = true;
+
+}
+
+void CreateIGCProgressDialog(const TCHAR* text) {
+   dlgIGCProgressShowModal();
+   IGCProgressDialogText(text);
+}
+
+
+void IGCProgressDialogText(const TCHAR* text) {
+  if(_WndForm)
+  {
+	 WindowControl* wText = _WndForm->FindByName(TEXT("frmIGCText"));
+    if(wText)
+    {
+        wText->SetCaption(text);
+        wText->Redraw();
+#ifndef USE_GDI
+        MainWindow.Refresh();
+#endif
+    }
+  }
+}
+
+
+
+ void OnIGCAbortClicked(WndButton* pWnd) {
+TCHAR Tmp[80 ];
+    (void)pWnd;
+
+	StartupStore(_T("Abort pressed%s"), NEWLINE);
+	_sntprintf(Tmp,80,  _T("%s ?"),MsgToken(2402));  // _@M2402_ "Abort Download"
+
```

Comment:
why question mark is not inside language token ?
moving question mark inside language token allow to remove  string copy Tmp buffer allocated in stack ...